File: icinga/plugin/check_eventdb.pl
#! /usr/bin/perl -w
#
# check_eventdb.pl - nagios plugin
#
# $Id: check_eventdb.pl 69 2009-10-26 12:31:45Z wpreston $
#
# Copyright (C) 2007 Gerd Mueller / Netways GmbH
# Copyright (C) 2008,2009 NETWAYS GmbH
# with additions by Rainer Brinkmoeller
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#

# Changes:
# 
# 2009-06-11: Martin Fuerstenau <mf_at_maerber.de>
#             - Added the ability to check program filed from the database. It can be
#               be delivered by all agents but wasn't used yet 
#             - Added $opt_dbuser and $opt_dbpasswd in "default values".
#               So there is no need to add a lot of users/password in the Nagios command
#               definition file
# 2010-09-26: Eric Lippmann <eric.lippmann@netways.de>
#             - Modified plugin to be compatible with icinga eventdb cronk

use strict;
use Getopt::Long;
use File::Basename;
use DBI;

use vars qw(
  $opt_msg
  $opt_host
  $opt_priority
  $opt_help
  $opt_db
  $opt_dbtable
  $opt_dbuser
  $opt_dbpasswd
  $opt_dbhost
  $opt_critical
  $opt_warning
  $opt_label
  $opt_type
  $opt_facility
  $opt_url
  $opt_maxage
  $opt_resetregexp
  $opt_program
  $opt_perfdata
  $dbh
  $sth
  $result
  $update
  $matches
  $message
  $last
  $startfrom
  @sql
  @params
  @urlparams
  $PROGNAME
  $VERSION
);

use subs qw(
	add_part
	myexit
	print_usage
);

$PROGNAME = basename($0);
$VERSION  = '1.5';

# default values
$opt_db       = 'icinga_web';
$opt_dbtable  = 'event_db_event';
$opt_dbhost   = 'localhost';
$opt_dbuser   = 'icinga_web';
$opt_dbpasswd = 'icinga_web';
$opt_host     = "";
$opt_msg      = "";
$opt_priority = "";
$opt_label    = "";
$opt_type     = "";
$opt_facility = "";
$opt_resetregexp     = "";
$opt_program = "";
$opt_perfdata = "";
$startfrom    = "0";
$opt_maxage   = '';

#GetOptions
Getopt::Long::Configure('bundling');
GetOptions(
	"msg=s"        => \$opt_msg,
	"m=s"          => \$opt_msg,
	"l=s"          => \$opt_label,
	"label=s"      => \$opt_label,
	"host=s"       => \$opt_host,
	"H=s"          => \$opt_host,
	"t=s"          => \$opt_type,
	"type=s"       => \$opt_type,
	"f=s"          => \$opt_facility,
	"facility=s"   => \$opt_facility,
	"priority=s"   => \$opt_priority,
	"p=s"          => \$opt_priority,
	"resetregexp=s"=> \$opt_resetregexp,
	"r=s"          => \$opt_resetregexp,
	"program=s"    => \$opt_program,
	"P=s"          => \$opt_program,
	"help"         => \$opt_help,
	"h"            => \$opt_help,
	"db=s"         => \$opt_db,
	"dbtable=s"    => \$opt_dbtable,
	"dbuser=s"     => \$opt_dbuser,
	"dbpassword=s" => \$opt_dbpasswd,
	"dbhost=s"     => \$opt_dbhost,
	"critical=i"   => \$opt_critical,
	"c=i"          => \$opt_critical,
	"warning=i"    => \$opt_warning,
	"w=i"          => \$opt_warning,
	"url=s"        => \$opt_url,
	"u=s"          => \$opt_url,
	"maxage=s"     => \$opt_maxage,
	"perfdata=s"   => \$opt_perfdata,
);

if ( !$opt_critical || !$opt_warning || $opt_help ) {
	print_usage();
}

# convert $opt_maxage, if defined
if ($opt_maxage ne '') {
	# first, make a timestamp
	my $timestamp = time();

	# then split integer from string
	my $int = (split(/(d|h|m)/, $opt_maxage))[0];

	# and subduct from timestamp
	$timestamp = $timestamp - $int * 86400 if $opt_maxage =~ /d/;	# days
	$timestamp = $timestamp - $int * 3600  if $opt_maxage =~ /h/;   # hours
	$timestamp = $timestamp - $int * 60    if $opt_maxage =~ /m/;   # minutes

	# format timestamp to mysql datetime
	my ($sec,$min,$hour,$mday,$mon,$year) = (localtime($timestamp))[0,1,2,3,4,5];
	$opt_maxage = sprintf("%02d-%02d-%02d %02d:%02d:%02d", $year+1900,$mon+1,$mday,$hour,$min,$sec);
}

# count from the last match
if ($opt_perfdata) {
	$opt_perfdata =~ /count=(\d*)/;
	$startfrom = $1 if ($1 and ($1 > 0));
}


# Connecting to Database
$dbh = DBI->connect( "DBI:mysql:database=$opt_db;host=$opt_dbhost",
	$opt_dbuser, $opt_dbpasswd, { PrintError => 0 } )
  || myexit( 'UNKNOWN', "can't connect to database!" );

# Count matches
push @sql, sprintf('SELECT COUNT(event_id) as count, MAX(event_id) as last from %1$s where event_id>%2$d', $opt_dbtable, $startfrom);

add_part('event_host', 1, $opt_host)
	if ($opt_host ne '');

add_part('event_message', 0, $opt_msg)
	if ($opt_msg ne '');

add_part('event_priority', 1, $opt_priority)
	if ($opt_priority ne '');

add_part('event_type', 1, $opt_type)
	if ($opt_type ne '');

add_part('event_facility', 1, $opt_facility)
	if ($opt_facility ne '');

add_part('event_created', 1, $opt_maxage, ">=")
	if ($opt_maxage ne '');

add_part('event_program', 0, $opt_program)
	if ($opt_program ne '');

add_part('event_ack', 0, '0');


# preparing the query
$sth = $dbh->prepare(join(' ', @sql))
  || myexit( 'UNKNOWN', "Can't prepare the query: ". $dbh->errstr );

# executing the query
$result = $sth->execute(@params)
  || myexit( 'UNKNOWN', "Can't execute the query: ". $sth->errstr );

$matches = 0;
$message = "";
if ($result) {
	while(my @row = $sth->fetchrow_array()) {
		($matches, $last) = @row;
		$message = $row[0];
	}

	$sth->finish();
}
if (!($last) or (!($last > 0))) {
	$last = $startfrom;
}


$dbh->disconnect()
  || myexit( 'UNKNOWN', "Can't disconnect: ". $dbh->errstr );;


if ( $matches >= $opt_critical ) {
    if ( $opt_resetregexp eq '' ) {
    	myexit( 'CRITICAL', $message, "matches=$matches count=$last"."c" );
    }
    elsif ( $message =~ m/($opt_resetregexp)/ ) {
    	myexit( 'OK', "Matches found already reseted.", "matches=$matches count=$last"."c" );
    }
    else {
    	myexit( 'CRITICAL', $message, "matches=$matches count=$last"."c" );
    }
	#myexit( 'CRITICAL', $matches . " matches found!", "matches=$matches count=$last"."c" );
}
elsif ( $matches >= $opt_warning ) {
    if ( $opt_resetregexp eq '' ) {
    	myexit( 'WARNING', $message, "matches=$matches count=$last"."c" );
    }
    elsif ( $message =~ m/($opt_resetregexp)/ ) {
    	myexit( 'OK', "Matches found already reseted.", "matches=$matches count=$last"."c" );
    }
    else {
    	myexit( 'WARNING', $message, "matches=$matches count=$last"."c" );
    }
	#myexit( 'WARNING', $matches . " matches found.", "matches=$matches count=$last"."c" );
}
else {
	myexit( 'OK', $matches . " matches found.", "matches=$matches count=$last"."c" );
}

myexit( 'UNKNOWN', 'Default exit' );

sub urlencode {
	my $str = "@_";
	$str =~ s/([^A-Za-z0-9])/sprintf("%%%02X", ord($1))/seg;
	return $str;
}

sub myexit {
	my $time;
	my $date;

	my ( $status, $text, $perfdata ) = @_;

	my %STATUS_CODE =
	  ( 'UNKNOWN' => '-1', 'OK' => '0', 'WARNING' => '1', 'CRITICAL' => '2' );

	my $out = undef;
	if ($opt_url) {
		$out .= sprintf('%1$s: %2$s %3$s', $status, $opt_label, 
		'<a href="' . $opt_url . '?' . join('&', @urlparams) . '">' . $text . '</a>');
	} else {	
		$out .= sprintf('%1$s: %2$s %3$s', $status, $opt_label, $text);
	}	
	$out .= sprintf('|%1$s', $perfdata) if ($perfdata);
	$out .= "\nmessage filter: $opt_msg";
	$out .= "\nreset regexp: $opt_resetregexp";
	$out .= chr(10);

	print $out;

	exit $STATUS_CODE{$status};
}

sub add_part {
	my ($field, $is_array, $value, $op, $agg) = @_;

	$agg = 'AND' if (!defined $agg);
	$op = '=' if (!defined $op);

	if ($value =~ m/\*|\%/ && lc($op) ne 'regexp') {
		$op = 'LIKE';
		$value =~ s/\*/\%/g;
	}

	push @sql, sprintf(
		'%1$s %2$s %3$s ?',
		$agg, $field, $op
	);

	if ($is_array == 1) {$field .= '[]'}
	push @urlparams, "$field=" . urlencode($value);
	push @params, $value;

	return 1;
}

sub print_usage {
	print <<EOU;
    Usage: $PROGNAME [ -H host ] [ -p priority ] [ -t type ] [ -m msg ] [ -P program ] [ --db db ] [ --dbtabe dbtable ] [ --dbuser dbuser ] [ --dbpassword dbpassword ] [ --dbhost dbhost ] [ -l label ] [ --perfdata '\$SERVICEPERFDATA\$' ] [ --maxage STRING ] [ -r resetregexp ] -w warn -c crit

    Options:

    -H --host
        Hostname as logged by agent.
    -p --priority
        Priority as logged by agent.
    -m --msg
        Message as logged by agent (SQL Format wildcards).
    -t --type
    	The logtype (e.g. syslog, snmptrap, mail).
    -P --program
        Program as logged by agent.
    --db STRING
        Database (default: eventdb)
    --dbtable STRING
        Tablename (default: events)
    --dbuser STRING
        Databaseuser (default: none)
    --dbpassword STRING
        Databasepassword (default: none)
    --dbhost STRING
        Databaseserver (default: localhost)
    -l --label STRING
        label for plugin output
    -u --url STRING
        URL for EventDB link in plugin output
    --maxage STRING
        Max age of EventDB entry (eg. 1m, 2h, 3d)
    -r --resetregexp STRING
        Regular Expression for message entry in eventdb to change each state back to OK.
    --perfdata STRING
        performance data from the last check (e.g. \$SERVICEPERFDATA\$)
    -w --warning INTEGER
        number matches to result in warning status
    -c --critical INTEGER
        number of matches to result in critical status

EOU

	myexit( "UNKNOWN", $PROGNAME );
}


# define service{
#         use                             generic-service
#         host_name                       localhost
#         service_description             eventdb
#         # critical for all status other than 4.
#         check_command                   check_eventdb!1!1!%Interesting Status:%!-r Status: 4
#         }
# 'check_eventdb' command definition
# 
# define command{
#         command_name            check_eventdb
#         command_line            $USER1$/contrib/check_eventdb.pl --dbuser eventdbrw --dbpassword eventdbrw --url "/nagios/eventdb/index.php" -w $ARG1$ -c $ARG2$ -m "$ARG3$" "$ARG4$"
# }

File: plugin/eventdb/bin/check_eventdb.py
import struct
import socket
#! /usr/bin/python

from twisted.python.compat import inet_pton
import os.path

import getopt, pprint, sys, re, urllib
import eventdb

from eventdb.checkfilter import *
from eventdb.dbhandler import *
from optparse import OptionParser
from eventdb.connPoolDaemonProxy import *

class CheckStatusException(Exception):
    def __init__(self,status,output,perfdata = ""):
        self.status = status
        self.output = output
        self.perfdata = perfdata


class EventDBPlugin(object):
    def __init__(self,arguments = None,noExit = False):
        self.__noExit = noExit;
        self.__requestStrategies = [DBHandler()]
        if arguments == None:
            self.__options = self.__parseArguments()
        elif isinstance(arguments,object):
            self.__options = arguments;
  
        self.__prepareArguments()
        self.__runCheck()

  
    def __prepareArguments(self):
        self.__handleArguments()
        if(self.__options.print_cv):
            print self.__getCVFilter()
            return
        try:
            self.__validateArguments()


        except Exception, e:
            self.__pluginExit("UNKNOWN","Invalid Arguments",e)


    def __runCheck(self):
        options = self.__options
        result = False

        try:
            if options.daemon_pid != False:
                self.__requestStrategies.insert(
                    0,
                    ConnPoolDaemonProxy(
                        options.daemon_pid,
                        options.daemon_log,
                        options.daemon_behaviour,
                        options.daemon_lifetime
                    )
                )
            if result == False :
                result = self.__dbQuery()
            if(result):
                self.__checkResult(result[2],result[3],result[0],result[1],result[4])

        except CheckStatusException, cs:
            raise
        except SystemExit, s:
            raise
        except Exception, e:
            self.__pluginExit("UNKNOWN","An error occured",e)


    def __handleArguments(self):
        options = self.__options
        
        self.__checkFilter = CheckFilter()
        self.__checkFilter.setLogtype(options.logtype)
        self.__checkFilter.setMaxage(options.maxage)
        self.__checkFilter.setPerfdata(options.perfdata)
        self.__checkFilter.setFacility(options.facility)
        self.__checkFilter.setPriority(options.priority)
        self.__checkFilter.program = options.program
        self.__checkFilter.message = options.message
        self.__checkFilter.host = options.host
        self.__checkFilter.ipaddress = options.ipaddress
        if(self.__options.facility):
            self.__options.facility = self.__options.facility.split(",")
        if(self.__options.priority):
            self.__options.priority = self.__options.priority.split(",")



    def __validateArguments(self):
        if(self.__options.warning == -1 or self.__options.critical == -1):
            raise Exception("warning or critical parameter missing")

    '''
    '   Returns the CustomVariable definition required to add special eventDB filter to the icinga-web
    '   cronk
    '
    '''
    def __getCVFilter(self):
        opts = self.__options
        strTpl = "_edb_filter            {'msg': %s, 'sourceExclusion': [%s],'priorityExclusion': [%s], 'facilityExclusion': [%s], 'startTime': %s }"
        msgFilter = "''"
        sourceExclusion = ""
        priorityExclusion = ""
        facilityExclusion = ""
        timespan = -1
        if opts.message != "":
            msgFilter = "{type:'exact','message' : '%s', isRegexp: false}" % opts.message
        if opts.logtype != "":
            arr = ['0','1','2']
            del arr[opts.logtype]

            sourceExclusion = ",".join(arr)
        if isinstance(opts.priority,list):
            allPrios = ['0','1','2','3','4','5','6','7'];
            priorityExclusion = ",".join(list(set(allPrios)-set(opts.priority)))
        if isinstance(opts.facility,list):
            allFacs = [];
            for i in range(24):
                allFacs.append(str(i))

            facilityExclusion = ",".join(list(set(allFacs)-set(opts.facility)))
        if hasattr(opts,"start_ts") and opts.start_ts != '':
            timespan = opts.start_ts

        return strTpl % (msgFilter,sourceExclusion, priorityExclusion,facilityExclusion,timespan)


    '''
    '   Creates and returns an instance of DBHandler for the passed arguments
    '
    '''
    def __setupDB(self,strategy):
        db = strategy
        db.connect(
            self.__options.db_type,
            self.__options.db_host,
            self.__options.db_user,
            self.__options.db_password,
            self.__options.db_name,
            self.__options.db_port
        )
        return db
   

    def __dbQuery(self):
        for strategy in self.__requestStrategies:

            try :
                db = self.__setupDB(strategy)
                query = self.__buildQuery()

                cursor = db.execute(query)

                values = [0,0]

                
                for row in cursor:
                    if(len(row) != 4):
                        raise DatabaseException("SQL Query failed, returned wrong values")
                    values = [row[0],row[1],row[2],row[3]]
                
                if(values[1] == None):
                    values[1] = 0
                cursor = db.execute("SELECT message FROM %s WHERE id = %d" % (self.__options.db_table, values[1]))
                for row in cursor:
                    values.append(row[0])
                    return values

                self.__pluginExit('OK',"0 critcal and 0 warning matches found.\n","matches=0 count=%dc" % (self.__checkFilter.startfrom));

            except SystemExit, e:
                raise
            except CheckStatusException, cs:
                raise
            except Exception, e:
                if strategy == self.__requestStrategies[-1]:
                    self.__pluginExit('UNKNOWN', "",e)


    def __checkResult(self,warnings,criticals,count, last,msg = ""):

        #strip newlines from message
        if(msg != "" and isinstance(msg,str)):
            msg= msg.replace('\n',' ')
        if(criticals >= self.__options.critical):
            if(self.__options.resetregex and re.search(self.__options.resetregex,msg)):
                return self.__pluginExit(
                    'OK',
                    "%d critical and %d warning matches found\nMatches found already reseted." % (criticals,warnings),
                    'matches=%d count=%dc' % (count,last)
                )
            else:
                return self.__pluginExit(
                    'CRITICAL',
                    ("%d critical and %d warning matches found\n"+msg) % (criticals,warnings),
                    'matches=%d count=%dc' % (count,last)
                )
        elif(warnings >= self.__options.warning):
            if(self.__options.resetregex  and re.search(self.__options.resetregex,msg)):
                return self.__pluginExit(
                    'OK',
                    "%d critical and %d warning matches found\nMatches found already reseted."% (criticals,warnings),
                    'matches=%d count=%dc' % (count,last)
                )
            else:
                return self.__pluginExit(
                    'WARNING',
                    ('%d critical and %d warning matches found \n,'+msg) % (criticals,warnings),
                    'matches=%d count=%dc' % (count,last)
                )
        else:
            return self.__pluginExit(
                'OK',
                "%d critical and %d warning matches found."%(criticals,warnings),
                "matches=%d count=%dc"%(count,last)
            )

        return self.__pluginExit('UNKNOWN', '0 matches found.\n','Default exit')



    def __buildQuery(self):
        o = self.__options

        filter = self.__checkFilter
        warningCount = self.__getCountField(o.prio_warning,"count_warning");
        criticalCount = self.__getCountField(o.prio_critical,"count_critical");
        queryBase = "SELECT COUNT(id) as count, MAX(id) as last , "+warningCount+", "+criticalCount+" from %s where id > %d "
       
        queryBase = queryBase % (o.db_table,filter.startfrom)
        urlParams = {}
        if(filter.host != ""):
            queryBase += self.__getWherePart("upper(host_name)",filter.host.upper())
            urlParams["host[]"] = filter.host

        if(filter.message != ""):
            queryBase += self.__getWherePart("message",filter.message)
            urlParams["message"] = filter.message

        if(filter.priority != ""):
            queryBase += self.__getWherePart("priority",filter.priority)
            urlParams["priority[]"] = filter.priority

        if(filter.facility != ""):
            queryBase += self.__getWherePart("facility",filter.facility)
            urlParams["facility[]"] = filter.facility

        if(filter.logtype != ""):
            queryBase += self.__getWherePart("type",filter.logtype)
            urlParams["type[]"] = filter.logtype

        if(filter.maxage != ""):
            queryBase += self.__getWherePart("created",filter.maxage, ">=")
            urlParams["created[]"] = filter.maxage

        if(filter.program != ""):
            queryBase += self.__getWherePart("program",filter.program)
            urlParams["program[]"] = filter.program

        if(filter.ipaddress != ""):
            queryBase += self.__getWherePart("host_address",self.__convertIp(filter.ipaddress))
            urlParams["address[]"] = filter.ipaddress

        queryBase += self.__getWherePart("ack",0)
        urlParams["ack"] = "0"
        o.urlParams = urlParams

        return queryBase;
    
    def __convertIp(self,address):
        p = re.compile("\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}")
        if p.search(address) != None :
            address = "::ffff:"+address
        
        address = inet_pton(socket.AF_INET6, address)
        
        if self.__options.db_type == "oracle":
            imedAddress = ""
            for byte in address :
                imedAddress += "%0.2X" % ord(byte)
            print imedAddress
            return imedAddress
        return address


    def __getCountField(self,values, field):
        if(values == ""):
            return " COUNT(id) AS "+field+" "
        caseQuery = "CASE WHEN PRIORITY IN ("+values+") THEN 1 ELSE 0 END"
        return " SUM("+caseQuery+") AS "+field+" "


    def __getWherePart(self,field,value,op = "=",agg = "AND"):
        
        if(isinstance(value,list)):
            op = "IN"
            value = "("+",".join(value)+")"
        else:
            value = str(value)
            if(re.search(r"\*|\%",value) != None):
                op = "LIKE"
                value = re.sub(r"(\*)","%",value)
        tpl = agg+" "+field+" "+op+" "
        if(value.isdigit()):
            value = str(int(value))
        elif op != 'IN':
            value = "'"+value+"' "
        else:
            value = value+" "
        tpl += value

        return " "+tpl+" "


    def __pluginExit(self,status,text,perfdata):
        statusCode = {'UNKNOWN' : 3, 'OK' : 0, 'WARNING' : 1, 'CRITICAL' : 2}
        out = ""
        if(self.__options.url != "" and hasattr(self.__options,"urlParams")):
            params = urllib.urlencode(self.__options.urlParams);
            link = '<a href="%s?%s">%s</a>' % (self.__options.url,params,text)
            out += "%s: %s %s" % (status, self.__options.label, link)
        else:
            out += "%s %s %s" % (status,self.__options.label,text)

        out += '|%s' % perfdata
        out += "\nmessage filter: %s" % self.__options.message
        out += "\nreset regexp: %s" % self.__options.resetregex
        out += chr(10)
        
        if self.__noExit == False:
            print out
            sys.exit(statusCode[status])
        raise CheckStatusException(statusCode[status],str(text),str(perfdata))

        

    def __parseArguments(self):
        parser = OptionParser()
        parser.add_option("-H","--host",dest="host",
                        help="Hostname as logged by the agent", default="")
        parser.add_option("-m","--msg",dest="message",
                        help="Message as logged by the agent (SQL Format wildcards)", default="")
        parser.add_option("-f","--facility",dest="facility", default="",
                        help="The facilities to respect, comma separated"),
        parser.add_option("-p","--priority",dest="priority",
                        help="Priority as logged by the agent", default="")
        parser.add_option("-t","--type",dest="logtype",type="choice",default="syslog",
                        help="The logtype (syslog,snmptrap,mail)",choices=["syslog","snmptrap","mail"])
        parser.add_option("-P","--program",dest="program",
                        help="Program as logged by the agent", default="")
        parser.add_option("-W","--warning-priorities",dest="prio_warning",default="",
                        help="A comma seperated set of priorities which will be used for determine the warning state" ),
        parser.add_option("-C","--critical-priorities",dest="prio_critical",default="",
                        help="A comma seperated set of priorities which will be used for determine the warning state" ),
        parser.add_option("--db",dest="db_name",
                        help="Name of the database to query",
                        default="eventdb")
        parser.add_option("--dbtype",dest="db_type",type="choice",
                        help="Type of the database to query (mysql or oracle)",
                        default="mysql",choices=["mysql","oracle"])
        parser.add_option("--dbtable",dest="db_table",
                        help="Name of the database table (usually event)",
                        default="event")
        parser.add_option("--dbuser",dest="db_user",
                        help="User for db login",default="eventdb")
        parser.add_option("--dbpassword",dest="db_password",
                        help="Password for db login",default="eventdb")
        parser.add_option("--dbhost",dest="db_host",
                        help="DB Host", default="localhost")
        parser.add_option("--dbport",dest="db_port",type="int",
                        help="DB Port", default=None)
        parser.add_option("-l","--label",dest="label",
                        help="Label for plugin output", default="")
        parser.add_option("-u","--url",dest="url",
                        help="URL for EventDB link in plugin output", default="")
        parser.add_option("--maxage",dest="maxage",
                        help="Maximum age of EventDB entry (eg. 1m, 2h, 3d)", default="")
        parser.add_option("-r","--resetregexp",dest="resetregex",
                        help="Regular Expression for message entry in eventdb to change each state back to OK", default="")
        parser.add_option("--perfdata",dest="perfdata",
                        help="Performance data from the last check (e.g. \$SERVICEPERFDATA\$)", default="")
        parser.add_option("--ip",dest="ipaddress", help="Filter by ip address", default="")
        parser.add_option("-w","--warning",dest="warning",type="int",help="Number of matches to result in warning state",default="-1")
        parser.add_option("-c","--critical",dest="critical",type="int",help="Number of matches to result in critical state",default="-1")
        parser.add_option("--cventry",dest="print_cv", default=False,action="store_true",
                        help="returns the custom variable entry for this call (needed in order to use icinga-web cronk integration)")
        parser.add_option("--daemon_pid", dest="daemon_pid", default=False,
                        help="Location of eventdb_daemon.pid which will be used for connection pooling. If no daemon is currently running, the plugin will spawn one.")
        parser.add_option("--daemon_log", dest="daemon_log", default=DEFAULT_LOG)
        parser.add_option("--daemon_behaviour", dest="daemon_behaviour", choices=["Aggressive","Servile"], default="Aggressive",
                        help="Defines how this daemon behaves when another daemon is started at (exactly) the same time."+
                        "(Aggressive[=Default]: Other daemons will be removed"+
                        ",Servile: Abort daemonization when another daemon is encountered)")
        parser.add_option("--daemon_lifetime", dest="daemon_lifetime", default=5,type="int",
                        help="Defines how long daemonized connections life (in seconds) when there's no request")
        (options, args) = parser.parse_args()
        return options;


  


def main():
    EventDBPlugin()

if __name__ == "__main__":
    main()

# define service{
#         use                             generic-service
#         host_name                       localhost
#         service_description             eventdb
#          _edb_filter                     {..result From cventry..}   
#           # critical for all status other than 4.
#         check_command                   check_eventdb!1!1!%Interesting Status:%!-r Status: 4
#         }
# 'check_eventdb' command definition
# 
# define command{
#         command_name            check_eventdb
#         command_line            $USER1$/contrib/check_eventdb.pl --dbuser eventdbrw --dbpassword eventdbrw --url "/nagios/eventdb/index.php" -w $ARG1$ -c $ARG2$ -m "$ARG3$" "$ARG4$"
# }

